test/unit/Cardano/Wallet/DB/MVarSpec.hs

```diff
@@ -71,44 +112,84 @@ dbReplaceValsProp db (key, val1, val2)  = monadicIO $ liftIO $ do
     putCheckpoint db key (initWallet val1)
     putCheckpoint db key (initWallet val2)
     resFromDb <- readCheckpoint db key
-
     resFromDb `shouldBe` (Just $ initWallet val2)
 
 dbMultiplePutsSeqProp
     :: DBLayer IO DummyState
-    -> KeyValPairs
+    -> KeyValPairs (PrimaryKey WalletId) DummyState
     -> Property
 dbMultiplePutsSeqProp db (KeyValPairs keyValPairs) = monadicIO $ liftIO $ do
     mapM_ (\(key, val) -> putCheckpoint db key (initWallet val)) keyValPairs
     resFromDb <- Set.fromList <$> readWallets db
-
     resFromDb `shouldBe` (Set.fromList (map fst keyValPairs))
 
 dbMultiplePutsParProp
     :: DBLayer IO DummyState
-    -> KeyValPairs
+    -> KeyValPairs (PrimaryKey WalletId) DummyState
     -> Property
 dbMultiplePutsParProp db (KeyValPairs keyValPairs) = monadicIO $ liftIO $ do
     mapConcurrently_
         (\(key, val) -> putCheckpoint db key (initWallet val))
         keyValPairs
     resFromDb <- Set.fromList <$> readWallets db
-
     resFromDb `shouldBe` (Set.fromList (map fst keyValPairs))
 
+dbMergeTxHistoryProp
+    :: DBLayer IO DummyState
+    -> KeyValPairs (PrimaryKey WalletId) (Map (Hash "Tx") (Tx, TxMeta))
+    -> Property
+dbMergeTxHistoryProp db (KeyValPairs keyValPairs) =
+    cover 90 cond "conflicting tx histories" prop
+  where
+    restrictTo k = filter ((== k) . fst)
+    -- Make sure that we have some conflicting insertion to actually test the
+    -- semantic of the DB Layer.
+    cond = map sumOfSizes ids /= map sizeOfUnion ids
```

Comment:
so in 90% of cases we have the situation that the same key occurred more than once in keyValPairs and hence the values coalesced. If yes, can't we just put :  
```
(L.length . L.nub) ids < L.length ids
``` 
?

Reply:
Hmmm.... :thinking: It's true that we only get conflict if we have the same wallet ids and some similar tx ids. So it should be enough indeed to just look at the ids and not at the balance to know whether we're going to have conflict. The balance being different is just a consequence of the id being in conflict. Good point. I'll simplify that :+1: